t/FHEM/lib/SD_Protocols/03_ProtocolData.t

```diff
@@ -4,22 +4,15 @@ use warnings;
 
 use Test2::V0;
 use Test2::Tools::Compare qw{is item U D match hash array bag};
-use Mock::Sub;
 use Test2::Todo;
 
-use lib::SD_Protocols
+plan(2);
+my $todo = Test2::Todo->new(reason => 'added to TEST 03_ProtocolData.t');
+if (defined($todo)) {
```

Comment:
Mir ist nicht klar, was dieser Test hier machen soll.

Reply:
*smile*

DAS ist KEIN TEST.
Das ist die Vorlage wo du dich frei entfalten sollst / kannst :-)

Das ToDo kommt wieder heraus.
- In der Datei muss eine Schleife laufen von allen Protokollen.
- Wo der Eintrag Modulation vorhanden ist, dort muss geprüft werden ob der Eintrag register existiert.
- Im nächsten Schritt muss der Wert welcher im register steht geprüft werden ob dieser so ist. Grund dafür, wenn jemand diesen ändert, so muss ein Fehler auftreten weil wir nur mit den Werten sicher stellen, das der Empfang auch funktioniert.

Kurzum, ich habe eine Arbeitsanweisung geschaffen *grins*

Reply:
Dass diese Test noch nichts testet hatte ich verstanden.

Schleife auf alle Protokolle verstehe ich.
Wenn es das Attribut Modulation gibt soll geprüft werden ob es auch ein Attribut Register gibt.
Auch das soweit verstanden.

Gehen was soll aber nun das Attribut Register geprüft werden? Das habe ich noch nicht verstanden.

Reply:
zur Richtigstellung:

>Wenn es das Attribut Modulation gibt soll geprüft werden ob es auch ein Attribut Register gibt.
Auch das soweit verstanden.

Nicht Attribute, sondern der Schlüssel register / modulation der ProtokollID.
Die bisherigen Register müssen gegen den derzeitigen Wert geprüft werden wie dieser in der ProtokollDatei steht derzeit.
Nur diese Einstellung des Register garantiert den Empfang / später das senden. Wenn man diesen Wert ausverstehn verstellt und wir bekommen das nicht mit, so suchen wir den Wolf.

Reply:
Ich habe mich schlecht ausgedrückt.
Schlüssel hatte ich gemeint :)

Wo befinden sich die bisherigen Register?
Wo befindet sich die derzeitigen Werte?

Reply:
https://github.com/HomeAutoUser/RFFHEM/blob/dev-r35_xFSK_TEST/FHEM/lib/SD_ProtocolData.pm#L2730

Hier startend.

Reply:
Ich Versuche es anders zu formulieren,
Gegen was soll ich in der lib denn den Wert register vergleichen?

Reply:
```
    "100" =>  # Lacrosse, Mode 1 - IT+

        register        => ['0001','0246','0302','042D','05D4','06FF','0700','0802','0D21','0E65','0F6A','1089','115C','1206','1322','14F8','1556','1700','1818','1916','1B43','1C68','1D91','23EC','2517','2611','2B3E'],

    "101" =>  # ELV PCA 301

        register        => ['0001','0246','0307','042D','05D4','06FF','0700','0802','0D21','0E6B','0FD0','1088','110B','1206','1322','14F8','1553','1700','1818','1916','1B43','1C68','1D91','23ED','2517','2611','2B3E'],

    "102" =>  # KoppFreeControl

        register        => ['0001','012E','0246','0304','04AA','0554','060F','07E0','0800','0900','0A00','0B06','0C00','0D21','0E65','0F6A','1097','1183','1216','1363','14B9','1547','1607','170C','1829','1936','1A6C','1B07','1C40','1D91','1E87','1F6B','20F8','2156','2211','23EF','240A','253D','261F','2741'],

    "103" =>  # Lacrosse Mode 2 - IT+

        register        => ['0001','0246','0302','042D','05D4','06FF','0700','0802','0D21','0E65','0F6A','1088','1182','1206','1322','14F8','1556','1700','1818','1916','1B43','1C68','1D91','23EC','2516','2611','2B3E'],

    "104" =>  # Remote control TR60C-1 with touch screen from Satellite Electronic (Zhongshan) Ltd., Importer Westinghouse Lighting for ceiling fan Bendan
```

Sobald diese Register in der `SD_ProtocolData.pm` verändert werden, so muss der Test anschlagen.
Grund, nur diese Einstellung sichert den Empfang, da sonst der CC1101 falsch ggf eingestellt wird.
Die Werte müssen soit überwacht werden, das dort nicht mal ausversehen sich was verändert oder diese verschwinden.

Reply:
Okay, ich habe es jetzt verstanden.

Du möchtest nicht testen ob der Code funktioniert, es ist eher eine Art Daten Verifizierung.

Möchte jemand was anpassen an den Registern, muss er es dann an zwei Stellen tun.

Das können wir so machen.
Wobei sich das auch irgendwie mit unseren Referenzdaten aus dem SIGNALduino Tool überschneidet :)

Reply:
Schön wenn du es verstanden hast :-)

> Wobei sich das auch irgendwie mit unseren Referenzdaten aus dem SIGNALduino Tool überschneidet :)

Der Test überschneidet sich nicht mit dem SIGNALduino_TOOL Test. Wir testen dort nur die RAW Message welche wir dispatchen und die dann entstandenen Readingwerte.

Die Register an sich, testen wir da nicht und deswegen empfand ich den Test als sehr Sinnvoll und zur Absicherung der Funktion. Registerwerte vom cc1101 sind Einstellungen der Hardware um erstmal auch die Chance zu erhalten zu empfangen oder zu senden. Dies ist eine etwas "aufwendigere Sache" als bei normalen ASK/OOK.